Add support for using cluster name for the delete cluster command

axbarsan opened the pull request:
Towards https://github.com/giantswarm/gsctl/issues/188

At the moment, this only provides support for the delete cluster command
- Renamed `ClusterIDMissing` errors to `ClusterNameOrIDMissing`
- Before executing the delete request, it gets all the clusters and checks to see if there is a cluster with the provided ID or name in there
- Changed error messages to also say cluster **name**, not just **ID**

When a cluster is deleted, the confirmation will state the name/ID provided by the user. So if the user provided an ID, the confirmation will show the ID; or a name, if he provided a name.

**Preview**
* Deleting a cluster with the name
![image](https://user-images.githubusercontent.com/13508038/75463528-3ea96500-5986-11ea-887d-0f61680ac7cd.png)

* Deleting a cluster with the ID
![image](https://user-images.githubusercontent.com/13508038/75463649-64366e80-5986-11ea-9d93-0cd760de736d.png)

* Trying to delete a cluster with an ID/name that doesn't exist
![image](https://user-images.githubusercontent.com/13508038/75463717-79130200-5986-11ea-80da-50bb508197cc.png)

* Trying to delete a cluster that has the same name as another existing one
![image](https://user-images.githubusercontent.com/13508038/75863904-80f8f900-5e01-11ea-8847-882972f5e2ba.png)

* Saving cluster IDs in the cache file (`~/.config/gsctl/clustercache.yaml`)
![image](https://user-images.githubusercontent.com/13508038/75777676-a977e900-5d56-11ea-93e5-c06918b86681.png)

As a side note, this only includes the `delete cluster` command, because I want to know that my implementation is right, before copying and pasting it to the other commands.
Also, I would like to do 1-2 commands per PR, to keep them small, so we could prevent any unwanted bugs.

axbarsan committed:
Add support for using the cluster name for the delete cluster command

axbarsan committed:
Confirm deletion with the cluster name

axbarsan committed:
Add support for multi-word confirm texts

ubergesundheit reviewed (approved):
Cool!

oponder commented on commands/delete/cluster/command.go:
```diff
@@ -185,32 +186,63 @@ func printResult(cmd *cobra.Command, args []string) {
+	}
+
+	for _, cluster := range response.Payload {
+		if cluster.ID == clusterNameOrID || cluster.Name == clusterNameOrID {
```
Here we basically go for the first found match right?

I think since names do not have to be unique, we should fail with a "Cluster name is ambiguous" type of failure message when the name you chose is present more than once.

It would be pretty rare, but I can imagine someone naming their clusters "Dev" "Production" and "Staging" and then having multiple organisations with those clusters all with the same name or something like that. So it's not too far fetched.

axbarsan replied:
Ahhh yesss, I actually had this in my notes but completely forgot about this

Thank you for noticing, I'll update and throw an error if there are multiple clusters with the same names

marians replied:
Yep, even if rare, it would be fatal, so I consider that absolutely necessary.

oponder reviewed (changes requested):
See my comment on the name ambiguity edge case. Might need a bit of UX consideration in how we present that message. 

Like the coolest would be to offer some kind of picker that is like 
```
Found more than one cluster called "My Cluster", please type the id of the cluster that you would like to delete:

CLUSTER ID    ORGANIZATION    NAME
123ab         testing         My Cluster 
456cd         production      My Cluster 
```

But if that needs more UX consideration, then for now just failing and asking the user to use a cluster ID instead would be ok with me.

axbarsan commented:
> See my comment on the name ambiguity edge case. Might need a bit of UX consideration in how we present that message.
> 
> Like the coolest would be to offer some kind of picker that is like
> 
> ```
> Found more than one cluster called "My Cluster", please type the id of the cluster that you would like to delete:
> 
> CLUSTER ID    ORGANIZATION    NAME
> 123ab         testing         My Cluster 
> 456cd         production      My Cluster 
> ```
> 
> But if that needs more UX consideration, then for now just failing and asking the user to use a cluster ID instead would be ok with me.

I like that, I'll try to implement it

marians commented:
So my understanding is we basically have to list clusters in every command that needs a cluster ID argument currently, before we can issue the API call. Because we don't know in an input like

    gsctl create kubeconfig -c a1b2c

whether `a1b2c` is a cluster ID or cluster name.

In installations with many organizations and clusters we see this taking several seconds (in one case, more than 5). So every command would get delayed by that duration. To me, that is too much of a downside.

What are other ways we can think of? Some ideas:

- Maintain a cache of last seen cluster IDs: If input `a1b2c` is a hit in our cache, we assume that it's an ID and don't try to resolve name to ID first. (We could re-build that cache periodically and/or whenever we call `GET /v4/clusters/`)
- Allow users to turn off the feature? When turned off, the behaviour should be just like it is currently.

axbarsan commented:
Yes, we do fetch all clusters in every command.

I believe caching wouldn't be an ideal solution, especially if we encounter situations like this:
- Creating a cluster in `gsctl`
- Deleting it through `happa` or direct API call and creating a new one with the same name
- Deleting it in `gsct` would throw an error, because in the cache there's another ID stored for that name, and the API will error out, saying it does not exist

> Allow users to turn off the feature? When turned off, the behaviour should be just like it is currently.
I would vote for this

Also, would it make sense to have a `-n|--name` param that will specify the name? So it will only make an extra API call if you do it that way

marians commented:
The caching idea was not about caching cluster names. Rather only cache cluster IDs we really got from the API. So every hit on that cache could be considered a valid cluster ID that at least _has_ existed at some point.

axbarsan commented:
Oh I get it now, so we'll cache the cluster IDs, so if we provided an ID, it would check the cache and if it's there, it would not make the API call. That sounds good, I can start on that

axbarsan committed:
Handle name collisions and print existing clusters with same name in a table

axbarsan committed:
Move cluster name/ID resolver to a separate file

axbarsan committed:
Move table printing to separate function

axbarsan committed:
Write tests for the 'GetClusterByID' function

axbarsan committed:
Replace colliding cluster ID in delete cluster command

axbarsan commented:
I added support for name collisions, which also prints a table with the colliding clusters, and then you can enter the ID of the cluster that you want to use.

Updated the description with the preview.

Next: Adding the ID cache

axbarsan committed:
Create cluster cacher

axbarsan committed:
Add comments to caching functions

axbarsan committed:
Fix tests

axbarsan committed:
Add tests for cluster caching

axbarsan commented:
Also added caching and unit tests for caching

axbarsan committed:
Cache all clusters into cache in the 'GetClusterByID()' function

marians commented:
After I opened the  `util` package, I learned that it isn't good Go practice to have such a catch-all package. In Go, devs expect explanatory package names. Look at all the folders on the top level. It would fit the current layout of the repo to just add a package like `cache` or, more descriptive, `cluster-cache`.

marians commented on util/cluster.go:
```diff
@@ -0,0 +1,205 @@
+// GetClusterID gets the cluster ID for a provided name/ID
+// by checking in both the user cache and on the API
+func GetClusterID(clusterNameOrID string, clientWrapper *client.Wrapper) (string, error) {
+	isID := IsInClusterCache(clusterNameOrID)
```
"isID" implies that we can find out whether this is an ID. In fact, we can only find out if it is an ID we know. So `isKnownID` or `isCachedID` would fit better.

marians commented on util/cluster.go:
```diff
@@ -0,0 +1,205 @@
+
+	default:
+		return clusterIDs[0], nil
+	}
```
I think an if / else if construct and a simple return after that would make this easier to read, no?

marians commented on util/cluster.go:
```diff
@@ -0,0 +1,205 @@
+		e, _ := readClusterCache(fs)
+		existingC <- e
+	}()
+	existing := <-existingC
```
Why use a channel to read a file's content?

marians replied:
And why read the existing cache content at all? When you cache the last cached items, why do you care what was in the cache before? This will grow the list forever, no?

axbarsan replied:
My initial thinking was that we might call this method in different places, not just when fetching all clusters. But now that you pointed it out, it probably won't get used in other places, so it doesn't really make sense to append to the existing cache (which in the first case, it should've had a limit)

axbarsan replied:
And about the channels, I wanted to somehow make cache writing non-blocking to the actual commands that use it, but I think I'm missing something in the actual implementation, because it doesn't to work like that

marians commented on util/cluster.go:
```diff
@@ -0,0 +1,205 @@
+		return []string{}, err
+	}
+
+	c := strings.Split(string(output), ",")
```
The separator should be a constant in the package. BTW, I would prefer a newline. That would make the file much more handy if one should ever have to debug something in a text editor.

marians commented on util/cluster.go:
```diff
@@ -0,0 +1,205 @@
+
+func writeClusterCache(fs afero.Fs, c ...string) error {
+	filePath := path.Join(config.ConfigDirPath, clusterCacheFileName)
+	output := []byte(strings.Join(c, ","))
```
As above, please use a separator constant

marians commented:
I haven't looked at everything yet, so might have missed it. One more questions regarding the cluster cache: Does the cache ever expire? Would cluster IDs get re-fetched when the cache is older than a certain duration?

One thing to keep in mind is that the cache is not endpoint/installation specific. So what is a valid ID in one installation might not be one in another. But with cluster IDs as random as they are, I don't see that as a problem.

axbarsan commented:
Thank you for the thorough review, @marians . I really appreciate the suggestions you've made.

I didn't take into consideration expiry dates and endpoint-specific IDs, but I think it may be worth switching to a YAML structure to introduce those, to make it adding other functionality easier to do in the future

axbarsan committed:
Improve code style by renaming variables and cutting unnecessary switch blocks

axbarsan committed:
Move cluster caching utils to separate package

axbarsan committed:
Convert cache format to YAML

- Save cache as YAML
- Implement separate cluster ID caches per endpoint
- Implement expiration for each endpoint cluster IDs

axbarsan committed:
Fix clustercache tests

axbarsan committed:
Change clustercache file extension to .yaml

axbarsan commented:
* Made the cache save to a YAML file
* Added different IDs per API endpoint
* Added expiration dates for cluster caches

Updated the description with a preview of the file

axbarsan committed:
Add comments and clean up code

marians commented:
Nitpick: Some people in the company will be much happier to find sentences in comments ending in a full stop. Do as you wish.

marians commented on clustercache/clustercache.go:
```diff
@@ -0,0 +1,240 @@
+	} else if len(matchingIDs) > 1 {
+		// There are multiple IDs that correspond to that cluster name
+		// Help the user decide which cluster to pick
+		_, id := handleNameCollision(clusterNameOrID, response.Payload)
```
If we don't need the first return value from `handleNameCollision`, I think we should remove it from the functions signature.

axbarsan committed:
Update comments to have periods at the end

marians commented on clustercache/clustercache.go:
```diff
@@ -0,0 +1,240 @@
+	}
+
+	// Write the cache file
+	_ = write(fs, cache)
```
Same as above: if the return value of `write` is not interesting, then rather make it not return anything.

axbarsan committed:
Remove first return member of 'handleNameCollision' return tuple

axbarsan committed:
Return from ID cacher if there is an error

axbarsan committed:
Actually declare the error variable

marians reviewed (approved):
We spoke about a little UX improvement to be made when the name is ambiguous.

Besides that, great work! ❤️

axbarsan committed:
Include release date and create date in the name collision table

axbarsan commented:
Removed the `Name` column from the name collision table. Added `Release` and `Created ` columns instead.
Updated the description with the preview

axbarsan committed:
Update the name collision messages
